File: Year-2023/LeetCode-May-Challenge/Dota2Senate.java
import java.util.*;
class Solution {
    public String predictPartyVictory(String senate) {
        int n = senate.length();
        Queue<Integer> radiant = new LinkedList<>(), dire = new LinkedList<>();
        for(int i = 0; i < n; ++i){
            if(senate.charAt(i) == 'R') radiant.add(i);
            else dire.add(i);
        }
        if(radiant.isEmpty()) return "Dire";
        else if(dire.isEmpty()) return "Radiant";
        while(!radiant.isEmpty() && !dire.isEmpty()){
            if(radiant.peek() < dire.peek()){
                dire.remove();
                int rad = radiant.poll();
                radiant.add(n + rad);
            } else{
                radiant.remove();
                int dir = dire.poll();
                dire.add(n + dir);
            }
        }
        return radiant.isEmpty() ? "Dire": "Radiant";
    }
}
public class Dota2Senate{
    public static void main(String[] args) {
        String senate = "RD";
        System.out.println(new Solution().predictPartyVictory(senate));
    }
}

File: Year-2023/LeetCode-May-Challenge/Dota2Senate.js
var predictPartyVictory = function (senate) {
    let n = senate.length;
    let radiant = [], dire = [];
    for (let i = 0; i < n; ++i) {
        if (senate[i] === 'R') radiant.push(i);
        else dire.push(i);
    }
    if (radiant.length === 0) return "Dire";
    else if (dire.length === 0) return "Radiant";
    let i = n;
    while (radiant.length && dire.length) {
        let rad = radiant.shift(), dir = dire.shift();
        if (rad < dir) radiant.push(i++);
        else dire.push(i++);
    }
    return radiant.length ? "Radiant" : "Dire";
};

let senate = "RD";
console.log(predictPartyVictory(senate));